test-framework/src/main/java/org/apache/brooklyn/test/framework/TestSshCommandImpl.java

```diff
@@ -0,0 +1,269 @@
+
+        Result result = null;
+        String downloadUrl = getConfig(DOWNLOAD_URL);
+        String command = getConfig(COMMAND);
+
+        String downloadName = DOWNLOAD_URL.getName();
+        String commandName = COMMAND.getName();
+
+        Map<String, Object> env = getConfig(SHELL_ENVIRONMENT);
+        if (env == null) env = ImmutableMap.of();
+        
+        if (!(isNonBlank(downloadUrl) ^ isNonBlank(command))) {
+            throw illegal("Must specify exactly one of", downloadName, "and", commandName);
+        }
+
+        if (isNonBlank(downloadUrl)) {
+            String scriptDir = getConfig(SCRIPT_DIR);
+            String scriptPath = calculateDestPath(downloadUrl, scriptDir);
+            result = executeDownloadedScript(machineLocation, downloadUrl, scriptPath, env);
+        }
+
+        if (isNonBlank(command)) {
+            result = executeShellCommand(machineLocation, command, env);
+        }
+
+        handle(result);
+    }
+
+    private Result executeDownloadedScript(SshMachineLocation machineLocation, String url, String scriptPath, Map<String, Object> env) {
+
+        TaskFactory<?> install = SshTasks.installFromUrl(ImmutableMap.<String, Object>of(), machineLocation, url, scriptPath);
+        DynamicTasks.queue(install);
+        DynamicTasks.waitForLast();
+
+        List<String> commands = ImmutableList.<String>builder()
+                .add("chmod u+x " + scriptPath)
+                .addAll(maybeCdToRunDirCmd())
+                .add(scriptPath)
+                .build();
+
+        return runCommands(machineLocation, commands, env);
+    }
+
+    private Result executeShellCommand(SshMachineLocation machineLocation, String command, Map<String, Object> env) {
+
+        List<String> commands = ImmutableList.<String>builder()
+                .addAll(maybeCdToRunDirCmd())
+                .add(command)
+                .build();
+
+        return runCommands(machineLocation, commands, env);
+    }
+
+    private List<String> maybeCdToRunDirCmd() {
+        String runDir = getConfig(RUN_DIR);
+        if (!isBlank(runDir)) {
+            return ImmutableList.of(CD + " " + runDir);
+        } else {
+            return ImmutableList.of();
+        }
+    }
+
+    private Result runCommands(SshMachineLocation machine, List<String> commands, Map<String, Object> env) {
+        @SuppressWarnings({ "unchecked", "rawtypes" })
+        SshEffectorTasks.SshEffectorTaskFactory<Integer> etf = SshEffectorTasks.ssh(commands.toArray(new String[]{}))
+                .environmentVariables((Map<String, String>)(Map)env)
+                .machine(machine);
+
+        ProcessTaskWrapper<Integer> job = DynamicTasks.queue(etf);
+        job.asTask().blockUntilEnded();
+        return new Result(job);
+    }
+
+
+
+    private IllegalArgumentException illegal(String message, String... messages) {
+        Iterable<String> allmsgs = Iterables.concat(MutableList.of(this.toString() + ":", message), Arrays.asList(messages));
```

Comment:
Just for interest, why change this from the previous implementation that just joined the arrays?